### PR TITLE
Saved view proposals

### PR DESCRIPTION
**Pull request by benjaminpkane**
### Changes

* Removes `changing_saved_view` and `update` from state events
* Initializes routing and state through initial `getEventSource` connects
* Ensures `router` is used to avoid desktop issues
* Removes `ipython` specific checks
* Moves saved views fragment to the Dataset route query
* Adds the `queryParams` parameter to routing and the `view` search parameter for the Dataset route

### CODE REVIEW

File: app/packages/state/src/hooks/useStateUpdate.ts [hunk 1]
```diff
@@ -52,6 +52,9 @@ const useStateUpdate = () => {
 
       if (state) {
         const view = get(viewAtoms.view);
+        if (dataset.stages && !state.view) {
+          state.view = dataset.stages;
```
Comment (kaixi-wang) [inline app/packages/state/src/hooks/useStateUpdate.ts:56]: Does this mean that If dataset.stages exist, then we are loading a saved view? 

In that case, would it make sense or be possible to set state.view_name here?

File: app/packages/core/src/Dataset.ts [hunk 2]
```diff
@@ -8,13 +8,33 @@ import {
...
+
 export const DatasetNodeQuery = graphql`
   query DatasetQuery(
     $name: String!
     $view: BSONArray = null
     $viewName: String = null
   ) {
+    ...DatasetSavedViewsFragment
     dataset(name: $name, view: $view, viewName: $viewName) {
+      stages(slug: $viewName)
```
Comment (kaixi-wang) [inline app/packages/core/src/Dataset.ts:37]: should `viewName` be `slug`? or vice versa? I think you use slug below.

File: app/packages/state/src/hooks/useStateUpdate.ts (hunk 1 above)
Comment (benjaminpkane) [reply app/packages/state/src/hooks/useStateUpdate.ts:55]: Isn't `viewName` already set from routing? Is there specific behavior you are thinking about?

File: app/packages/core/src/Dataset.ts (hunk 2 above)
Comment (benjaminpkane) [reply app/packages/core/src/Dataset.ts]: Resolved, I believe. Thanks!

File: fiftyone/server/events.py [hunk 3]
```diff
@@ -256,18 +256,53 @@ async def _initialize_listener(payload: ListenPayload) -> InitializedListener:
...
-            state.changing_saved_view = False
+                state.selected = []
+                state.selected_labels = []
+                state.view = None
+                state.view_name = None
+            else:
+                if payload.initializer.view:
+                    try:
+                        state.view = state.dataset.load_saved_view(
+                            payload.initializer.view
```
Comment (kaixi-wang) [inline fiftyone/server/events.py:283]: is the `payload.initializer` the url? and so would `payload.initializer.view` be the value of `view` in the search params?

File: fiftyone/server/events.py [hunk 4]
```diff
@@ -256,18 +256,53 @@ async def _initialize_listener(payload: ListenPayload) -> InitializedListener:
...
+                        )
+                        state.selected = []
+                        state.selected_labels = []
+                        state.view_name = payload.initializer.view
+                    except:
+                        pass
+        elif (
+            payload.initializer.view
+            and payload.initializer.view != current_saved_view
```
Comment (kaixi-wang) [inline fiftyone/server/events.py:292]: if `payload.initializer.view` is from the url, its value is`view.slug`. Is `state.view_name` always going to be the slug? or only set here? If so, maybe rename it to slug to prevent setting it elsewhere in the code using `view.name`

File: fiftyone/server/view.py [hunk 5]
```diff
@@ -161,7 +158,7 @@ def get_saved_view_stages(identifier):
     return list(
         db.views.aggregate(
             [
-                {"$match": {"slug": identifier}},
+                {"$match": {"name": identifier}},
```
Comment (kaixi-wang) [inline fiftyone/server/view.py:161]: the `name` is only unique relative to a dataset. to identify a specific view, either have to use `_id` or `name`+ `_dataset_id`

File: fiftyone/core/dataset.py [hunk 6]
```diff
@@ -3137,14 +3137,13 @@ def save_view(
             raise ValueError("Cannot save view into a different dataset")
 
         view._set_name(name)
-        slug = self._validate_saved_view_name(name, overwrite=overwrite)
+        self._validate_saved_view_name(name, overwrite=overwrite)
```
Comment (kaixi-wang) [inline fiftyone/core/dataset.py:3140]: why are we removing slug?

File: fiftyone/core/dataset.py [hunk 7]
```diff
@@ -3301,13 +3296,11 @@ def _get_saved_view_doc(self, name, pop=False):
...
 
     def _validate_saved_view_name(self, name, skip=None, overwrite=False):
-        slug = fou.to_slug(name)
-
         for view_doc in self._doc.saved_views:
             if view_doc is skip:
                 continue
 
-            if name == view_doc.name or slug == view_doc.slug:
+            if name == view_doc.name:
```
Comment (kaixi-wang) [inline fiftyone/core/dataset.py:3303]: I don't' think this will work. Unique names will resolve to the same slug 
image

**Review by kaixi-wang — CHANGES_REQUESTED** [review]
I only looked at the python so far, but I don't think it's safe to just remove slug.

**Comment by benjaminpkane** [comment]
Still reviewing with respect to encoding and view names

File: app/packages/state/src/hooks/useStateUpdate.ts (hunk 1 above)
Comment (kaixi-wang) [reply app/packages/state/src/hooks/useStateUpdate.ts:55]: It wasn't previously. The view name is initially set in `set_view` and afterwards the url is updated in order to sync the other tabs.

File: fiftyone/core/dataset.py (hunk 6 above)
Comment (benjaminpkane) [reply fiftyone/core/dataset.py:3140]: I don't see the point of slug if it is not something made apparent to the user? What is the benefit?

File: fiftyone/core/dataset.py (hunk 7 above)
Comment (benjaminpkane) [reply fiftyone/core/dataset.py:3303]: Can you clarify? What is the issue?

File: fiftyone/server/events.py (hunk 4 above)
Comment (benjaminpkane) [reply fiftyone/server/events.py:292]: It will come from the URL but will have already been decoded via [decodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent). The server only needs to know the name.

**Comment by kaixi-wang** [comment]
haven't gotten a chance to test this branch and so not sure if your changes affect this, but if the view changes after someone loads a saved view (ie add/remove items from the view bar), it is no longer a saved view so the view selector should show unsaved view and the view param should be removed from the url.

File: fiftyone/server/view.py (hunk 5 above)
Comment (benjaminpkane) [reply fiftyone/server/view.py:161]: I don't think the function is used anywhere. Can I remove?

File: fiftyone/server/events.py (hunk 3 above)
Comment (benjaminpkane) [reply fiftyone/server/events.py:283]: This is a dataclass. It is defined [here](https://github.com/voxel51/fiftyone/blob/43a8752b3c48e86c7c9da8e9be05173ddf084adc/fiftyone/core/session/events.py#L98)

**Comment by benjaminpkane** [comment]
I added encoding/decoding of view in the URL. Still need to test edge cases and notebooks in the morning.

File: app/packages/relay/src/mutations/setView.ts [hunk 8]
```diff
@@ -6,8 +6,6 @@ export default graphql`
     $session: String
     $view: BSONArray!
     $viewName: String
-    $savedViewSlug: String
-    $changingSavedView: Boolean
```
Comment (manivoxel51) [inline app/packages/relay/src/mutations/setView.ts:10]: 👍 much simpler without these two

File: fiftyone/server/view.py (hunk 5 above)
Comment (kaixi-wang) [reply fiftyone/server/view.py]: yeah it's no longer used because directly querying the db like this isn't more efficient that iterating on the saved views property on the dataset singleton. You can remove it.

**Review by manivoxel51 — APPROVED** [review]
Thank you for your work @benjaminpkane ! 💯 

I am testing this locally

File: app/packages/core/src/Dataset.ts [hunk 9]
```diff
@@ -8,13 +8,33 @@ import {
...
+  fragment DatasetSavedViewsFragment on Query
+  @refetchable(queryName: "DatasetSavedViewsFragmentQuery") {
+    savedViews(datasetName: $name) {
+      id
+      datasetId
+      name
+      slug
+      description
+      color
+      viewStages
```
Comment (kaixi-wang) [inline app/packages/core/src/Dataset.ts:21]: Since we don't actually set the view using `viewStages`, we can probably remove this from the fragment to avoid passing around potentially really long stage definitions that are not viewed.

File: app/packages/state/src/utils.ts [hunk 10]
```diff
@@ -118,11 +119,11 @@ export type ResponseFrom<TQuery extends { response: unknown }> =
 export const getSavedViewName = (context: RoutingContext<any>): string => {
   const datasetName = getDatasetName(context);
   if (datasetName) {
-    const queryString = window.location.search;
+    const queryString = context.history.location.search;
     const params = new URLSearchParams(queryString);
     const viewName = params.get("view");
     if (viewName) {
-      return viewName;
+      return decodeURIComponent(viewName);
```
Comment (kaixi-wang) [inline app/packages/state/src/utils.ts:126]: is `getSavedViewName` used where the name specifically (not the slug) is required? If so, then this might not work. For example if `viewName = "saved view"`, then `slug = "saved-view"`, and`decodeURIComponent(viewName)==slug` but `decodeURIComponent(viewName)!=viewName`

**Review by manivoxel51 — COMMENTED** [review]
The selection doesn't seem to change when you click on them (even though the url is changing)

Screen Shot 2023-01-12 at 11 34 35 AM

**Review by manivoxel51 — COMMENTED** [review]
tested this in browser, desktop, notebook and found 2 issues. the first issue seem to be related to slug vs. name changes and the fix is 

- The view selection is always changing to default - even if after you select a view

https://user-images.githubusercontent.com/109545780/212212119-c72bd2e0-ea3f-48f8-bcbd-181f8f721ae0.mov

I found that the issue is savedViewParam is being set to view's name/label where the equality to find the selected is based on slug so label !== slug nothing selected. note this works for single worded views

- Accessing a saved view directly is broken via view=VIEW_SLUG. might be related to slug changes too

https://user-images.githubusercontent.com/109545780/212212310-edcdfc06-ecc6-4ab8-803d-7f2e23f1e795.mov

no solution yet

File: app/packages/core/src/components/Sidebar/ViewSelection/index.tsx [hunk 11]
```diff
@@ -51,71 +45,63 @@ export type DatasetViewOption = Pick<
...
-  const fragments = usePreloadedQuery(DatasetSavedViewsQuery, queryRef);
-  const [data, refetch] = useRefetchableFragment(
-    DatasetSavedViewsFragment,
-    fragments
-  );
+  const fragmentRef = useContext(DatasetQueryRef);
 
-  const items =
-    (data as { savedViews: fos.State.SavedView[] })?.savedViews || [];
+  if (!fragmentRef) throw new Error("ref not defined");
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/core/src/components/Sidebar/ViewSelection/index.tsx:71]: It wouldn't. But it asserts the type

File: app/packages/app/src/index.tsx [hunk 12]
```diff
@@ -162,7 +145,10 @@ const App: React.FC = ({}) => {
       },
       controller.signal,
       {
-        initializer: getDatasetName(contextRef.current),
+        initializer: {
+          dataset: getDatasetName(contextRef.current),
+          view: getSavedViewName(contextRef.current),
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/app/src/index.tsx:150]: This inspects the URL and initializes the view slug through the event source

File: app/packages/state/src/utils.ts (hunk 10 above)
Comment (benjaminpkane) [reply app/packages/state/src/utils.ts:126]: This method was always assuming the slug, but I will rename.

**Comment by benjaminpkane** [comment]
I fixed the `savedViewParam` issue. Thanks

I am not seeing any other issues, though.

**Comment by manivoxel51** [comment]
> I fixed the `savedViewParam` issue. Thanks
> 
> I am not seeing any other issues, though.

The direct access to the url with view=x is broken in **some** scenarios. the first half of the video shows loading a dataset and immediately trying to load view=A (when view B is loaded), the potentialView does not change and content is still content B (although view param in url is now A) - second half shows when it works and changes the view properly.


https://user-images.githubusercontent.com/109545780/212383269-0c9d9a88-3394-4bd1-9de2-1d6e01dc1d24.mov

**Comment by benjaminpkane** [comment]
This is a slug issue. I'll have to look at this weekend

File: app/packages/core/src/components/Sidebar/ViewSelection/index.tsx [hunk 13]
```diff
@@ -51,71 +45,63 @@ export type DatasetViewOption = Pick<
...
+  const searchData = useMemo(
     () =>
-      filter(
-        viewOptions,
-        ({ id, label, description, slug }: DatasetViewOption) =>
+      viewOptions.filter(
+        ({ id, label, description, slug }) =>
           id === DEFAULT_SELECTED.id ||
-          label.toLowerCase().includes(viewSearch) ||
+          label?.toLowerCase().includes(viewSearch) ||
```
[the remark this answers is not in the document]
Comment (kaixi-wang) [reply app/packages/core/src/components/Sidebar/ViewSelection/index.tsx:101]: is label the name?

**Review by kaixi-wang — APPROVED** [review]
just tested app, and everything looks good!
